runtime/src/main/java/io/quarkus/vault/runtime/VaultAuthManager.java

```diff
@@ -197,63 +202,39 @@ private String getPassword() {
-
-                try {
-                    unwrap = vaultInternalSystemBackend.unwrap(wrappingToken, clazz);
-                } catch (VaultClientException e) {
-                    if (e.getStatus() == 400) {
-                        String message = "wrapping token is not valid or does not exist; " +
-                                "this means that the token has already expired " +
-                                "(if so you can increase the ttl on the wrapping token) or " +
-                                "has been consumed by somebody else " +
-                                "(potentially indicating that the wrapping token has been stolen)";
-                        throw new VaultException(message, e);
-                    } else {
-                        throw e;
-                    }
-                }
-
-                wrappedValue = f.apply(unwrap);
-                wrappedCache.put(wrappingToken, wrappedValue);
-                String displayValue = getConfig().logConfidentialityLevel.maskWithTolerance(wrappedValue, LOW);
-                log.debug("unwrapped " + type + ": " + displayValue);
-                return wrappedValue;
-
-            } finally {
-                unwrapSem.release();
-            }
-        } catch (InterruptedException e) {
-            throw new RuntimeException("unable to enter critical section when unwrapping " + type, e);
-        }
-    }
-
-    private VaultKubernetesAuthAuth loginKubernetes() {
+    private <T> Uni<String> unwrapWrappingTokenOnce(String type, String wrappingToken, Function<T, String> f, Class<T> clazz) {
+        return unwrappingCache.computeIfAbsent(wrappingToken, (token) -> {
+            return vaultInternalSystemBackend.unwrap(wrappingToken, clazz)
+                    .map(unwrap -> {
+                        String wrappedValue = f.apply(unwrap);
+
+                        String displayValue = getConfig().logConfidentialityLevel.maskWithTolerance(wrappedValue, LOW);
+                        log.debug("unwrapped " + type + ": " + displayValue);
+
+                        return wrappedValue;
+                    })
+                    .onFailure(VaultClientException.class).transform(e -> {
+                        if (((VaultClientException) e).getStatus() == 400) {
+                            String message = "wrapping token is not valid or does not exist; " +
+                                    "this means that the token has already expired " +
+                                    "(if so you can increase the ttl on the wrapping token) or " +
+                                    "has been consumed by somebody else " +
+                                    "(potentially indicating that the wrapping token has been stolen)";
+                            return new VaultException(message, e);
+                        } else {
+                            return e;
+                        }
+                    })
+                    .memoize().indefinitely();
```

Comment:
this seems to be doing the trick yes. not a new an issue, but we are probably leaking memory here with that cache.

Reply:
Agreed. I will add a ticket to look at using an expiring cache.

Reply:
Added #51

Reply:
probably with `VaultDynamicCredentialsManager.credentialsCache` as well, although I doubt anybody will hit an OOM because of that one.